AG-13349 Fix dragging events

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-13349

olegat committed:
AG-13349 Capture mousemove / mouseup events when dragging

olegat committed:
AG-13349 Capture other mouse events too when dragging

olegat committed:
AG-13349 Dispatch mouseleave/mouseenter on drag-end when needed

olegat committed:
AG-13349 Do not update RegionManager.current when dragging

olegat committed:
AG-13349 Fix sonarjs/different-types-comparison error

olegat committed:
AG-13349 Only fire one startDrag once

This fixes a bug where two sets of window event handlers (capture) are running
when mousedown-ing an HTMLElement. For example, mousedown-ing a navigator handle
would fire 'drag-start' on the <input> slider and .ag-charts-canvas-container.

olegat committed:
AG-13349 Fix dragging on the series-area

olegat committed:
AG-13349 Migrate KeyNavManager to the widgets/ dragging mechanism

Also remove the dragging logic in InteractionManager which is no longer in use.

olegat committed:
AG-13349 Fix no-extra-boolean-cast eslint error

olegat committed:
AG-13349 Fix RegionEvent.canvasXY (fixes AG-12695 e2e test)

olegat committed:
AG-13349 Avoid multiple drag handlers using a static boolean

This fixes tests in zoom.test.ts

alantreadway commented on packages/ag-charts-community/src/widget/widgetListenerInternal.ts:
```diff
@@ -91,30 +98,62 @@ export class WidgetListenerInternal {
+        window.addEventListener('mouseout', mousegeneral, { capture: true });
+        window.addEventListener('mouseover', mousegeneral, { capture: true });
+        window.addEventListener('mousemove', mousemove, { capture: true });
+        window.addEventListener('mouseup', mouseup, { capture: true });
```
Bit of an edge case, but if the chart is destroyed in the middle of a drag, will these event handlers leak?

olegat replied:
No. How could that happen? `WidgetListenerInternal.destroy()` only removes `'mousedown'` event handlers on the HTMLElement, but if a drag is currently being processed then those `'mousedown'` event handles do not matter. The `'mouseup'` listener would still be registered in the `window` object, so it'll remove the handlers when it's called.

alantreadway replied:
>  so it'll remove the handlers when it's called.

The chart may have been destroyed before this point though, so there is a stale clousre still referencing the destroyed chart hanging around, and it would be very easy for that chart being in an unexpected state - to mean that these handlers fail in unexpected ways, before being removed.

Could you take a more defensive approach here, and ensure that on `destroy()` any of the temporary handlers are disposed of?

olegat replied:
I understand the importance of risk mitigation, but there are two important things to remember here:

1. We agreed on the design decision that the `widgets/` directory should be as isolated and independent as possible for separation of concerns (similar to the `scene/` directory). Therefore the state of the chart is of no relevance at this point. Only `this: WidgetListenerListener`, `window` and `current: T` (which is essentially an HTMLElement) are relevant.
2. The `this: WidgetListenerListener` is responsible for allocating these handlers but those not "own" them, they are conceptually owned by the `window`. Therefore removing the handlers in `WidgetListenerListener.destroy()` is more likely to cause problems than to solve anything. We would, for example, need to make sure that they are no bugs caused by double-destruction.

The only thing that the `mouseup` handler really needs from `this` is a callback. Therefore, we must go with an approach like this which is much much simpler (and safer):

```diff
diff --git a/packages/ag-charts-community/src/widget/widgetListenerInternal.ts b/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
index 164a36904..588f9f780 100644
--- a/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
+++ b/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
@@ -128,9 +128,8 @@ export class WidgetListenerInternal {
                 window.removeEventListener('mouseover', mousegeneral, { capture: true });
                 window.removeEventListener('mousemove', mousemove, { capture: true });
                 window.removeEventListener('mouseup', mouseup, { capture: true });
-                const dragEndEvent = makeDragEvent('drag-end', origin, upEvent);
-                this.dispatch('drag-end', current, dragEndEvent);
-                this.endDrag(current, dragEndEvent);
+                const callback: undefined | ((upEvent) => void) = getCallback(upEvent);
+                callback?.(upEvent);
             }
         };
 
```

olegat replied:
Fixed here: https://github.com/ag-grid/ag-charts/commit/a4709c65aa6f5dac828bf95b51b73c2c7c14a5f2

alantreadway replied:
Ok, let's roll with this then.

alantreadway commented on packages/ag-charts-community/src/chart/interaction/regionManager.ts:
```diff
@@ -261,12 +251,16 @@ export class RegionManager {
-                if (!this.isDragMoving) {
+            case 'leave':
+            case 'enter':
+                if (!this.isDragMoving || this.deferredDragStart != null) {
```
I'm not 100% following the logic here, but just to confirm that we won't issue drag-related events until a mouse move happens after a mouse down still, to avoid firing conflicting click and drag events on mouse-up?

olegat replied:
Correct. That's what the `blockNextClickEvent` state handles: We fire either 1 `'click'` event or a set of `'drag*'` events but not both.

The `case 'leave','enter':` logic stops us broadcasting which region changes when dragging. Example steps:

1. `'mousedown'` fired on the series-area.
2. `'mousemove'` fired such that the pointer leaves the bounds the series-area HTML element.
3. `'mouseleave'` fired as pointer hovers over a new HTML element (we don't want to broadcast this, because the user is still dragging).
4. `'mousemove'` fired such that the pointer returns to the bounds the series-area HTML element.
3. `'mouseenter'` fired as pointer hovers back to the series-area element (we don't want to broadcast this either; the user never really left because they are still dragging)

olegat committed:
AG-13349 CR: Use callback mechanism for drag events

This allows us to ignore drag events if WidgetListenerInternal.destroy() is called.

olegat committed:
AG-13349 Fix @typescript-eslint/no-shadow eslint error

alantreadway reviewed (approved):
LGTM
